Modified BCD class to be used as a computation class

AndrewAday opened the pull request:
Major changes:
    - BayesianChangepointDetection now extends Computation
    - Constructor takes istream and ostream arguments
    - Class will function as a computation, reading from istreams and
      producing the run-length probability array to ostreams
Note: end user will still have to manually call serve_computation, and provide
their own hazard and distribution.

AndrewAday committed:
Modified BCD class to be used as a computation class
Major changes:
	- BayesianChangepointDetection now extends Computation
	- Constructor takes istream and ostream arguments
	- Class will function as a computation, reading from istreams and
	  producing the run-length probability array to ostreams
Note: end user will still have to manually call serve_computation, and provide
their own hazard and distribution.

alanhdu commented on concord_ml/bcd/computations.py:
```diff
@@ -1,27 +1,64 @@
+# import logging
+# logging.basicConfig()
+# logger = logging.getLogger(__name__)
+# logger.setLevel(logging.DEBUG)
```
Could you get rid of the comments please?

alanhdu commented on concord_ml/bcd/computations.py:
```diff
@@ -1,27 +1,64 @@
 """Local implementation for Bayesian Changepoint Detection"""
-
 import numpy as np
+from concord.computation import (
```
This should probably be on one line

alanhdu commented on concord_ml/bcd/computations.py:
```diff
@@ -1,27 +1,64 @@
+            distribution of the current run length. Output[r_0] is the
+            probability that r=r_0 at time r_0.
+        """
+        result = self.step(float(record.data))
```
Hm... do we care what the key of the record is? It's fine if the answer's no.

alanhdu commented on concord_ml/bcd/computations.py:
```diff
@@ -1,27 +1,64 @@
+    """ Bayesian Changepoint Detection implementation
     Based on algorithm from http://arxiv.org/abs/0710.3742.
+    """
+    def __init__(self, hazard, distribution, istream=None, ostream=None):
```
`istream` really shouldn't ever be `None`. That'd be really silly.

I also feel like `ostream` shouldn't be `None` either, but who knows?

Also, if we're just going to pass in `[] if ostrema is None else ostream`, why don't we just set the default an empty tuple `()`?

alanhdu commented on concord_ml/bcd/computations.py:
```diff
@@ -1,27 +1,64 @@
+
+    def process_record(self, ctx, record):
+        """
+        Returns:
```
Could you update the docstring? We're not really returning anything anymore.

alanhdu commented on concord_ml/bcd/computations.py:
```diff
@@ -1,27 +1,64 @@
         self.time = 0
         self.Pr = np.ones(1)
 
+    def init(self, ctx):
```
Do we have to implement `init` if we don't do anything?

AndrewAday committed:
style changes, added support for isoformat record.key

process_record() now assumes record.key contains the isoformat timestamp
associated with the data. If the data is not ordered, i.e. it's timestamp
is older than the last data processed, we toss it.

AndrewAday committed:
updated travis test_bcd_against_reference()

Now passes istream and ostream into constructor for bcd computation

AndrewAday committed:
flake8 mods

alanhdu commented on concord_ml/bcd/computations.py:
```diff
@@ -1,26 +1,59 @@
+            result = self.step(float(record.data))
+            if self.ostream is not None:
+                for stream in self.ostream:
+                    ctx.produce_record(stream, '~', result)
```
Do we need to serialize this?

AndrewAday committed:
bcd process_record now sends argmax

AndrewAday committed:
removed whitespace
